modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -0,0 +1,321 @@
+        for (DataStream dataStream : state.metadata().dataStreams().values()) {
+            if (dataStream.getLifecycle() == null) {
+                continue;
+            }
+
+            IndexMetadata writeIndex = state.metadata().index(dataStream.getWriteIndex());
+            if (writeIndex != null && isManagedByDLM(dataStream, writeIndex)) {
+                RolloverRequest rolloverRequest = defaultRolloverRequestSupplier.apply(dataStream.getName());
+                transportActionsDeduplicator.executeOnce(
+                    rolloverRequest,
+                    ActionListener.noop(),
+                    (req, reqListener) -> rolloverDataStream(rolloverRequest, reqListener)
+                );
+            }
+
+            TimeValue retention = getRetentionConfiguration(dataStream);
+            if (retention != null) {
+                List<Index> backingIndices = dataStream.getIndices();
+                // we'll look at the current write index in the next run if it's rolled over (and not the write index anymore)
+                for (int i = 0; i < backingIndices.size() - 1; i++) {
+                    IndexMetadata backingIndex = state.metadata().index(backingIndices.get(i));
+                    if (backingIndex == null || isManagedByDLM(dataStream, backingIndex) == false) {
+                        continue;
+                    }
+
+                    TimeValue indexLifecycleDate = getCreationOrRolloverDate(dataStream.getName(), backingIndex);
+
+                    long nowMillis = nowSupplier.getAsLong();
+                    if (nowMillis >= indexLifecycleDate.getMillis() + retention.getMillis()) {
+                        // there's an opportunity here to batch the delete requests (i.e. delete 100 indices / request)
+                        // let's start simple and reevaluate
+                        DeleteIndexRequest deleteRequest = new DeleteIndexRequest(backingIndex.getIndex().getName()).masterNodeTimeout(
+                            TimeValue.MAX_VALUE
+                        );
+
+                        // time to delete the index
+                        transportActionsDeduplicator.executeOnce(
+                            deleteRequest,
+                            ActionListener.noop(),
+                            (req, reqListener) -> deleteIndex(deleteRequest, reqListener)
+                        );
+                    }
+                }
+            }
+        }
+    }
+
+    private void rolloverDataStream(RolloverRequest rolloverRequest, ActionListener<Void> listener) {
+        // "saving" the rollover target name here so we don't capture the entire request
+        String rolloverTarget = rolloverRequest.getRolloverTarget();
+        client.admin().indices().rolloverIndex(rolloverRequest, new ActionListener<>() {
+            @Override
+            public void onResponse(RolloverResponse rolloverResponse) {
+                listener.onResponse(null);
+            }
```

Comment:
Should we log at debug (or info?) level when DLM successfully rolled over an index? Seems like it would be very beneficial to have in the logs for tracking down supportability concerns.

Reply:
Yes, added logging for both rollover and retention (trace level when initiating the operation and info level on success)